packages/amplify-prompts/src/progressbars/progressbar.ts

```diff
@@ -0,0 +1,252 @@
+import { AmplifyTerminal as Terminal } from './terminal';
+
+/**
+ * Type for Bar configurations options
+ */
+export type BarOptions = {
+    // Custom formatter functions for main progress bar and item's underneath it
+    progressBarFormatter: (payload: ProgressPayload) => string,
+    itemFormatter: (payload: ItemPayload) => string,
+    // Indicates if the progress bar is part of a multiBar or standalone
+    loneWolf: boolean,
+    hideCursor: boolean,
+    // Specify characters for a completed/incomplete bar
+    barCompleteChar: string,
+    barIncompleteChar : string,
+    barSize: number,
+    // status which indicates item completion/failure
+    itemCompleteStatus: string[],
+    itemFailedStatus: string[],
+    // Text which appear on top of the progress bar indicating current status along with a spinner.
+    prefixText: string,
+    successText: string,
+    failureText: string
+}
+
+/**
+ * Type for main progress bar payload
+ */
+export type ProgressPayload = {
+    progressName: string,
+    envName: string
+}
+
+/**
+ * Type for progress bar item payload
+ */
+export type ItemPayload = {
+    LogicalResourceId: string,
+    ResourceType: string,
+    ResourceStatus: string,
+    TimeStamp: string,
+}
+
+type Item = {
+    name: string,
+    status: string,
+    renderString: string,
+    finished: boolean
+}
+
+/**
+ * Progress bar instance
+ */
+export class ProgressBar {
+    private value: number;
+    private total: number;
+    private terminal: Terminal;
+    private payload!: ProgressPayload;
+    private isActive: boolean;
+    private options: BarOptions;
+    private lastDrawnTime: number;
+    private items: Item[];
+
+    barCompleteString: string;
+    barIncompleteString: string;
+    barSize: number;
+
+    constructor(options: BarOptions) {
+      this.terminal = new Terminal();
+      this.value = 0;
+      this.total = 1;
+      this.options = options;
+      this.items = [];
+      this.isActive = false;
+      this.lastDrawnTime = 0;
+      this.barSize = options.barSize || 40;
```

Comment:
Where did `40` come from? Ideally it would be a constant with a descriptive name